File: source/cpp/83. Remove Duplicates from Sorted List.cpp
// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};


class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if (head == nullptr || head->next == nullptr) return head;

        ListNode* p = head;
        ListNode* temp = nullptr;
        while (p->next != nullptr) {
            if (p->val == p->next->val) {
                temp = p->next;
                p->next = p->next->next;
                delete temp;
            }
            else 
                p = p->next;
        }
        return head;
    }
};

File: source/python/83. Remove Duplicates from Sorted List.py
# Definition for singly-linked list.
from typing import Self


class ListNode:
    def __init__(self, val: int = 0, next: Self | None = None):
        self.val = val
        self.next: ListNode | None = next

    def __str__(self):
        return f"{self.val} -> {self.next}"

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, ListNode):
            return NotImplemented

        if other is None:
            return False

        return self.val == other.val and self.next == other.next


class Solution:
    def deleteDuplicates(self, head: ListNode | None) -> ListNode | None:
        if head is None:
            return None

        current = head
        while current.next is not None:
            if current.val == current.next.val:
                current.next = current.next.next
            else:
                current = current.next

        return head


if __name__ == '__main__':
    solution = Solution()

    assert solution.deleteDuplicates(
        ListNode(1, ListNode(1, ListNode(2)))) == ListNode(1, ListNode(2))
    assert solution.deleteDuplicates(ListNode(1, ListNode(1, ListNode(
        2, ListNode(3, ListNode(3)))))) == ListNode(1, ListNode(2, ListNode(3)))

    print("All tests passed!")
